rclcpp/src/rclcpp/executor.cpp

```diff
@@ -745,227 +593,133 @@ Executor::execute_client(
 }
 
 void
-Executor::wait_for_work(std::chrono::nanoseconds timeout)
-{
-  TRACETOOLS_TRACEPOINT(rclcpp_executor_wait_for_work, timeout.count());
-  {
-    std::lock_guard<std::mutex> guard(mutex_);
+Executor::collect_entities()
+{
+  // Get the current list of available waitables from the collector.
+  rclcpp::executors::ExecutorEntitiesCollection collection;
+  this->collector_.update_collections();
+  auto callback_groups = this->collector_.get_all_callback_groups();
+  rclcpp::executors::build_entities_collection(callback_groups, collection);
+
+  // Make a copy of notify waitable so we can continue to mutate the original
+  // one outside of the execute loop.
+  // This prevents the collection of guard conditions in the waitable from changing
+  // while we are waiting on it.
+  if (notify_waitable_) {
+    current_notify_waitable_ = std::make_shared<rclcpp::executors::ExecutorNotifyWaitable>(
```

Comment:
My initial thought was no, but I think it will need to be protected somehow.  We already have a mutex inside of ExecutorNotifyWaitable to protect the internals, I will likely reuse that.

Reply:
Added protection in the copy constructor of `ExecutorNotifyWaitable` in 8ed094e8